### ISSUE TITLE
cleanup build options for cmake

### ISSUE THREAD

**Issue by mathiaswagner** [lattice/quda#420]
We should mark some options as advanced if they are likely only to be changed by developers and not by users. It is also a good time to change names if desired.

**Comment by mathiaswagner** [comment]
my suggested sorting of options for now are shown below.
As cmake sorts options alphabetically I would prefer to rename all options for QUDA so that they start with QUDA (we can leave the defines the way they are, I am just thinking about the variable names in cmake).
Furthermore, if there are any options that can be unified or removed (preferably those that hardly affect compile time / library size) please comment now. I believe the switch to cmake is a great opportunity for that.

**Default options**
```
set(GPU_ARCH sm_35 CACHE STRING "set the GPU architecture (sm_20, sm_21, sm_30, sm_35, sm_50, sm_52)")

# build options
set(BUILD_WILSON_DIRAC ON CACHE BOOL "build Wilson Dirac operators")
set(BUILD_CLOVER_DIRAC ON CACHE BOOL "build clover Dirac operators")
set(BUILD_DOMAIN_WALL_DIRAC ON CACHE BOOL "build domain wall Dirac operators")
set(BUILD_STAGGERED_DIRAC ON CACHE BOOL "build staggered Dirac operators")
set(BUILD_TWISTED_MASS_DIRAC ON CACHE BOOL "build twisted mass Dirac operators")
set(BUILD_TWISTED_CLOVER_DIRAC ON CACHE BOOL "build twisted clover Dirac operators")
set(BUILD_NDEG_TWISTED_MASS_DIRAC OFF CACHE BOOL "build non-degenerate twisted mass Dirac operators")
set(BUILD_FATLINK OFF CACHE BOOL "build code for computing asqtad fat links")
set(BUILD_HISQLINK OFF CACHE BOOL "build code for computing hisq fat links")
set(BUILD_GAUGE_FORCE OFF CACHE BOOL "build code for (1-loop Symanzik) gauge force")
set(BUILD_FERMION_FORCE OFF CACHE BOOL "build code for asqtad fermion force")
set(BUILD_HISQ_FORCE OFF CACHE BOOL "build code for hisq fermion force")
set(BUILD_GAUGE_TOOLS OFF CACHE BOOL "build auxilary gauge-field tools")
set(BUILD_GAUGE_ALG OFF CACHE BOOL "build gauge-fixing and pure-gauge algorithms")
set(BUILD_CONTRACT OFF CACHE BOOL "build code for bilinear contraction")
# Dynamic inversion saves memory but decreases the flops
set(BUILD_DYNAMIC_CLOVER OFF CACHE BOOL "Dynamically invert the clover term for twisted-clover")
set(BUILD_QIO OFF CACHE BOOL "build QIO code for binary I/O")

# Multi-GPU options
set(BUILD_QMP OFF CACHE BOOL "set to 'yes' to build the QMP multi-GPU code")
set(BUILD_MPI OFF CACHE BOOL "set to 'yes' to build the MPI multi-GPU code")
set(POSIX_THREADS OFF CACHE BOOL "set to 'yes' to build pthread-enabled dslash")

#BLAS library
set(BUILD_MAGMA OFF CACHE BOOL "build magma interface")


# Interface options
set(BUILD_QDP_INTERFACE ON CACHE BOOL "build qdp interface")
set(BUILD_MILC_INTERFACE ON CACHE BOOL "build milc interface")
set(BUILD_CPS_INTERFACE OFF CACHE BOOL "build cps interface")
set(BUILD_QDPJIT_INTERFACE OFF CACHE BOOL "build qdpjit interface")
set(BUILD_BQCD_INTERFACE OFF CACHE BOOL "build bqcd interface")
set(BUILD_TIFR_INTERFACE OFF CACHE BOOL "build tifr interface")

# Packing option
set(QUDA_DEVICE_PACK ON CACHE BOOL "set to 'yes' to enable packing and unpacking on the device")

# QDPJIT
set(QUDA_USE_QDPJIT OFF CACHE BOOL "build QDP-JIT support?")

# Locations for QIO / QMP
set(QUDA_QIOHOME "" CACHE PATH "path to QIO")
set(QUDA_QMPHOME "" CACHE PATH "path to QMP")

```

**Advanced options**
```
set(QUDA_BLAS_TEX ON CACHE BOOL "enable texture reads in BLAS?")
set(QUDA_FERMI_DBLE_TEX ON CACHE BOOL "enable double-precision texture reads on Fermi?")
set(QUDA_NUMA_AFFINITY ON CACHE BOOL "enable NUMA affinity")
set(QUDA_VERBOSE_BUILD OFF CACHE BOOL "display kernel register useage")

# NVTX options
set(QUDA_MPI_NVTX OFF CACHE BOOL "add nvtx markup to MPI API calls for the visual profiler")
set(QUDA_INTERFACE_NVTX OFF CACHE BOOL "add nvtx markup to interface calls for the visual profiler")

# GPUdirect options
set(QUDA_GPU_DIRECT ON CACHE BOOL "set to 'yes' to allow GPU and NIC to shared pinned buffers")
set(QUDA_GPU_COMMS OFF CACHE BOOL "set to 'yes' to allow GPU and NIC to communicate using RDMA")

# features in development
set(BUILD_SSTEP OFF CACHE BOOL "build s-step linear solvers")
```

**Comment by mathiaswagner** [comment]
Just to clarify: Advanced options in cmake are only hidden in the GUI / text mode interface if the advanced option is not marked, see https://cmake.org/cmake/help/v3.0/command/mark_as_advanced.html

**Comment by maddyscientist** [comment]
This looks like a great idea.  Something like this would be much more readable given the alphabetical ordering :wink: 
```
set(GPU_ARCH sm_35 CACHE STRING "set the GPU architecture (sm_20, sm_21, sm_30, sm_35, sm_50, sm_52)")

# build options
set(BUILD_DIRAC_WILSON ON CACHE BOOL "build Wilson Dirac operators")
set(BUILD_DIRAC_CLOVER ON CACHE BOOL "build clover Dirac operators")
set(BUILD_DIRAC_DOMAIN_WALL ON CACHE BOOL "build domain wall Dirac operators")
set(BUILD_DIRAC_STAGGERED ON CACHE BOOL "build staggered Dirac operators")
set(BUILD_DIRAC_TWISTED_MASS ON CACHE BOOL "build twisted mass Dirac operators")
set(BUILD_DIRAC_TWISTED_CLOVER ON CACHE BOOL "build twisted clover Dirac operators")
set(BUILD_DIRAC_NDEG_TWISTED_MASS_DIRAC OFF CACHE BOOL "build non-degenerate twisted mass Dirac operators")
set(BUILD_FATLINK OFF CACHE BOOL "build code for computing asqtad fat links")
set(BUILD_HISQLINK OFF CACHE BOOL "build code for computing hisq fat links")
set(BUILD_FORCE_GAUGE OFF CACHE BOOL "build code for (1-loop Symanzik) gauge force")
set(BUILD_FORCE_FERMION OFF CACHE BOOL "build code for asqtad fermion force")
set(BUILD_FORCE_HISQ OFF CACHE BOOL "build code for hisq fermion force")
set(BUILD_GAUGE_TOOLS OFF CACHE BOOL "build auxilary gauge-field tools")
set(BUILD_GAUGE_ALG OFF CACHE BOOL "build gauge-fixing and pure-gauge algorithms")
set(BUILD_CONTRACT OFF CACHE BOOL "build code for bilinear contraction")
# Dynamic inversion saves memory but decreases the flops
set(BUILD_DYNAMIC_CLOVER OFF CACHE BOOL "Dynamically invert the clover term for twisted-clover")
set(BUILD_QIO OFF CACHE BOOL "build QIO code for binary I/O")

# Multi-GPU options
set(BUILD_QMP OFF CACHE BOOL "set to 'yes' to build the QMP multi-GPU code")
set(BUILD_MPI OFF CACHE BOOL "set to 'yes' to build the MPI multi-GPU code")
set(POSIX_THREADS OFF CACHE BOOL "set to 'yes' to build pthread-enabled dslash")

#BLAS library
set(BUILD_MAGMA OFF CACHE BOOL "build magma interface")


# Interface options
set(BUILD_INTERFACE_QDP ON CACHE BOOL "build qdp interface")
set(BUILD_INTERFACE_MILC ON CACHE BOOL "build milc interface")
set(BUILD_INTERFACE_CPS OFF CACHE BOOL "build cps interface")
set(BUILD_INTERFACE_QDPJIT OFF CACHE BOOL "build qdpjit interface")
set(BUILD_INTERFACE_BQCD OFF CACHE BOOL "build bqcd interface")
set(BUILD_INTERFACE_TIFR OFF CACHE BOOL "build tifr interface")

# Packing option
set(QUDA_DEVICE_PACK ON CACHE BOOL "set to 'yes' to enable packing and unpacking on the device")

# QDPJIT
set(QUDA_USE_QDPJIT OFF CACHE BOOL "build QDP-JIT support?")

# Locations for QIO / QMP
set(QUDA_QIOHOME "" CACHE PATH "path to QIO")
set(QUDA_QMPHOME "" CACHE PATH "path to QMP")
```

**Comment by mathiaswagner** [comment]
I'm also in favor of using QUDA as prefix for all our options. That keeps them together in the gui and CMakeCache, e.g.

QUDA_DIRAC_WILSON

etc.

Thoughts?

On 16.12.2015, at 17:52, maddyscientist <notifications@github.com<mailto:notifications@github.com>> wrote:


This looks like a great idea. Something like this would be much more readable given the alphabetical ordering [:wink:]

set(GPU_ARCH sm_35 CACHE STRING "set the GPU architecture (sm_20, sm_21, sm_30, sm_35, sm_50, sm_52)")

# build options
set(BUILD_DIRAC_WILSON ON CACHE BOOL "build Wilson Dirac operators")
set(BUILD_DIRAC_CLOVER ON CACHE BOOL "build clover Dirac operators")
set(BUILD_DIRAC_DOMAIN_WALL ON CACHE BOOL "build domain wall Dirac operators")
set(BUILD_DIRAC_STAGGERED ON CACHE BOOL "build staggered Dirac operators")
set(BUILD_DIRAC_TWISTED_MASS ON CACHE BOOL "build twisted mass Dirac operators")
set(BUILD_DIRAC_TWISTED_CLOVER ON CACHE BOOL "build twisted clover Dirac operators")
set(BUILD_DIRAC_NDEG_TWISTED_MASS_DIRAC OFF CACHE BOOL "build non-degenerate twisted mass Dirac operators")
set(BUILD_FATLINK OFF CACHE BOOL "build code for computing asqtad fat links")
set(BUILD_HISQLINK OFF CACHE BOOL "build code for computing hisq fat links")
set(BUILD_FORCE_GAUGE OFF CACHE BOOL "build code for (1-loop Symanzik) gauge force")
set(BUILD_FORCE_FERMION OFF CACHE BOOL "build code for asqtad fermion force")
set(BUILD_FORCE_HISQ OFF CACHE BOOL "build code for hisq fermion force")
set(BUILD_GAUGE_TOOLS OFF CACHE BOOL "build auxilary gauge-field tools")
set(BUILD_GAUGE_ALG OFF CACHE BOOL "build gauge-fixing and pure-gauge algorithms")
set(BUILD_CONTRACT OFF CACHE BOOL "build code for bilinear contraction")
# Dynamic inversion saves memory but decreases the flops
set(BUILD_DYNAMIC_CLOVER OFF CACHE BOOL "Dynamically invert the clover term for twisted-clover")
set(BUILD_QIO OFF CACHE BOOL "build QIO code for binary I/O")

# Multi-GPU options
set(BUILD_QMP OFF CACHE BOOL "set to 'yes' to build the QMP multi-GPU code")
set(BUILD_MPI OFF CACHE BOOL "set to 'yes' to build the MPI multi-GPU code")
set(POSIX_THREADS OFF CACHE BOOL "set to 'yes' to build pthread-enabled dslash")

#BLAS library
set(BUILD_MAGMA OFF CACHE BOOL "build magma interface")


# Interface options
set(BUILD_INTERFACE_QDP ON CACHE BOOL "build qdp interface")
set(BUILD_INTERFACE_MILC ON CACHE BOOL "build milc interface")
set(BUILD_INTERFACE_CPS OFF CACHE BOOL "build cps interface")
set(BUILD_INTERFACE_QDPJIT OFF CACHE BOOL "build qdpjit interface")
set(BUILD_INTERFACE_BQCD OFF CACHE BOOL "build bqcd interface")
set(BUILD_INTERFACE_TIFR OFF CACHE BOOL "build tifr interface")

# Packing option
set(QUDA_DEVICE_PACK ON CACHE BOOL "set to 'yes' to enable packing and unpacking on the device")

# QDPJIT
set(QUDA_USE_QDPJIT OFF CACHE BOOL "build QDP-JIT support?")

# Locations for QIO / QMP
set(QUDA_QIOHOME "" CACHE PATH "path to QIO")
set(QUDA_QMPHOME "" CACHE PATH "path to QMP")


-
Reply to this email directly or view it on GitHub<https://github.com/lattice/quda/issues/420#issuecomment-165170652>.

NVIDIA GmbH, Wuerselen, Germany, Amtsgericht Aachen, HRB 8361
Managing Director: Karen Theresa Burns

-----------------------------------------------------------------------------------
This email message is for the sole use of the intended recipient(s) and may contain
confidential information.  Any unauthorized review, use, disclosure or distribution
is prohibited.  If you are not the intended recipient, please contact the sender by
reply email and destroy all copies of the original message.
-----------------------------------------------------------------------------------

**Comment by maddyscientist** [comment]
Yes, sounds good.

**Comment by maddyscientist** [comment]
Closed with #423
